src/net/ossl.cc

```diff
@@ -673,196 +849,326 @@ class session : public enable_shared_from_this<session>, public session_impl {
         // encryption/decryption faster. Hence to avoid cases where we would do
         // an extra syscall for something like a 100 bytes header we linearize the
         // packet if it's below the max TLS record size.
-        // TODO(Rob): Avoid magic numbers
-        if (p.nr_frags() > 1 && p.len() <= 16000) {
+        if (p.nr_frags() > 1 && p.len() <= openssl_max_record_size) {
             p.linearize();
         }
-
-        auto i = p.fragments().begin();
-        auto e = p.fragments().end();
-        return with_semaphore(_out_sem, 1, std::bind(&session::do_put, this, i, e)).finally([p = std::move(p)] {});
+        return get_units(_out_sem, 1).then([this, p = std::move(p)](auto units) mutable {
+            return do_put(std::move(p), std::move(units)).then([this](net::packet p) {
```

Comment:
Units can be released within this lambda, doesn't need to be an argument to do_put